torchx/schedulers/kubernetes_mcad_scheduler.py

```diff
@@ -0,0 +1,1121 @@
+        V1Container,
+        V1ContainerPort,
+        V1EnvVar,
+        V1Pod,
+        V1PodSpec,
+        V1ResourceRequirements,
+        V1Service,
+    )
+    from kubernetes.client.rest import ApiException
+
+logger: logging.Logger = logging.getLogger(__name__)
+
+
+# Kubernetes reserves a small amount of resources per host for the system. For
+# TorchX we always assume the entire host is being requested so we adjust the
+# requested numbers account for the node reserved resources.
+#
+# https://kubernetes.io/docs/tasks/administer-cluster/reserve-compute-resources/
+RESERVED_MILLICPU = 100
+RESERVED_MEMMB = 1024
+
+RETRY_POLICIES: Mapping[str, Iterable[Mapping[str, str]]] = {
+    RetryPolicy.REPLICA: [],
+    RetryPolicy.APPLICATION: [
+        {"event": "PodEvicted", "action": "RestartJob"},
+        {"event": "PodFailed", "action": "RestartJob"},
+    ],
+}
+
+# The AppWrapper Status - holistic view of pods/services
+JOB_STATE: Dict[str, AppState] = {
+    # Pending is the AppWrapper condition waiting for scheduling by MCAD
+    "Pending": AppState.PENDING,
+    # Running is the AppWrapper condition in Running.
+    "Running": AppState.RUNNING,
+    # Deleted is the AppWrapper condition where the torchX job is cancelled
+    # by the user, the AppWrapper is deleted by the user, or error handling
+    "Deleted": AppState.CANCELLED,
+    # Failed is the job finishes unexpectedly
+    "Failed": AppState.FAILED,
+}
+
+TASK_STATE: Dict[str, ReplicaState] = {
+    # Pending dispatch means the AppWrapped task is not yet scheduled by MCAD
+    "Pending dispatch": ReplicaState.PENDING,
+    # Pending means the task is scheduled by MCAD
+    "pending": ReplicaState.PENDING,
+    # Running means a task is running on the host.
+    "running": ReplicaState.RUNNING,
+    # Succeeded means that all containers in the pod have voluntarily
+    # terminated with a container exit code of 0, and the system is not
+    # going to restart any of these containers.
+    "Succeeded": ReplicaState.SUCCEEDED,
+    # Failed means that all containers in the pod have terminated, and at
+    # least one container has terminated in a failure (exited with a
+    # non-zero exit code or was stopped by the system).
+    "failed": ReplicaState.FAILED,
+    # Unknown means the status of task/pod is unknown to the scheduler.
+    "Unknown": ReplicaState.UNKNOWN,
+}
+
+# TO DO update to standards
+LABEL_VERSION = "torchx.pytorch.org/version"
+LABEL_APP_NAME = "torchx.pytorch.org/app-name"
+LABEL_ROLE_INDEX = "torchx.pytorch.org/role-index"
+LABEL_ROLE_NAME = "torchx.pytorch.org/role-name"
+LABEL_REPLICA_ID = "torchx.pytorch.org/replica-id"
+
+ANNOTATION_ISTIO_SIDECAR = "sidecar.istio.io/inject"
+
+LABEL_INSTANCE_TYPE = "node.kubernetes.io/instance-type"
+
+
+def sanitize_for_serialization(obj: object) -> object:
+    from kubernetes import client
+
+    api = client.ApiClient()
+    return api.sanitize_for_serialization(obj)
+
+
+def role_to_pod(
```

Comment:
Is it possible to share this method between the kubernetes scheduler and mcad? I see the env differences, anything else?

Reply:
Primary differences are the function arguments (Volcano options or MCAD options) and the addition of the environment variable. This also leads to differences in the pod specs and metadata (section on V1Pod).